storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedHotBlockMigration.java

```diff
@@ -0,0 +1,98 @@
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.storage.server.kvstore.dataaccess.KvStoreCombinedDaoBlinded;
+import tech.pegasys.teku.storage.server.kvstore.dataaccess.KvStoreCombinedDaoUnblinded;
+
+public class BlindedHotBlockMigration<
+    T extends KvStoreCombinedDaoBlinded & KvStoreCombinedDaoUnblinded> {
+  private static final Logger LOG = LogManager.getLogger();
+  private static final int BATCH_SIZE = 1_000;
+  private final Spec spec;
+
+  private final T dao;
+
+  private BlindedHotBlockMigration(final Spec spec, final T dao) {
+    this.spec = spec;
+    this.dao = dao;
+  }
+
+  static <T extends KvStoreCombinedDaoBlinded & KvStoreCombinedDaoUnblinded> void migrateBlocks(
+      final T dao, final Spec spec) {
+    final BlindedHotBlockMigration<T> blindedHotBlockMigration =
+        new BlindedHotBlockMigration<>(spec, dao);
+    blindedHotBlockMigration.performBatchMigration();
+  }
+
+  private void performBatchMigration() {
+    moveHotBlocksToBlindedStorage();
+  }
+
+  private void moveFirstFinalizedBlockToBlindedStorage() {
+    Optional<SignedBeaconBlock> maybeBlock = dao.getEarliestFinalizedBlock();
+    maybeBlock.ifPresent(
+        block -> {
+          final Bytes32 root = block.getRoot();
+          LOG.info("Setting lowest finalized block at {}({})", root, block.getSlot());
+          try (KvStoreCombinedDaoBlinded.FinalizedUpdaterBlinded blindedUpdater =
+                  dao.finalizedUpdaterBlinded();
+              KvStoreCombinedDaoUnblinded.FinalizedUpdaterUnblinded unblindedUpdater =
+                  dao.finalizedUpdaterUnblinded()) {
+
+            blindedUpdater.addBlindedFinalizedBlock(block, root, spec);
+            unblindedUpdater.deleteUnblindedFinalizedBlock(block.getSlot(), root);
+            blindedUpdater.commit();
+            unblindedUpdater.commit();
+          }
+        });
+  }
+
+  private void moveHotBlocksToBlindedStorage() {
+    final long countBlocks = dao.countUnblindedHotBlocks();
+    if (countBlocks == 0) {
+      return;
+    }
+    LOG.info("Migrating blocks to blinded storage, {} hot blocks to migrate", countBlocks);
+
+    long counter = 0;
+    try (final Stream<SignedBeaconBlock> blocks = dao.streamHotBlocks()) {
+      for (Iterator<SignedBeaconBlock> it = blocks.iterator(); it.hasNext(); ) {
+        try (KvStoreCombinedDaoBlinded.FinalizedUpdaterBlinded blindedUpdater =
+                dao.finalizedUpdaterBlinded();
+            KvStoreCombinedDaoUnblinded.HotUpdaterUnblinded unblindedUpdater =
+                dao.hotUpdaterUnblinded()) {
+          for (int i = 0; i < BATCH_SIZE && it.hasNext(); i++) {
+            final SignedBeaconBlock block = it.next();
+            blindedUpdater.addBlindedBlock(block, block.getRoot(), spec);
+            unblindedUpdater.deleteUnblindedHotBlockOnly(block.getRoot());
+            counter++;
```

Comment:
Just in case this winds up happening in a long period of non-finality - it may be worth printing a progress message every 1000 blocks or something like that. Will almost certainly not be needed but will be a lot better than a very long period of silence if we ever do hit this case.